Get first card quicker during reviews

Arthur-Milchior opened the pull request:
Getting the first card when you select a deck is a long process (at least when you have a big deck collection). This is because before actually getting the first card, AnkiDroid computes the number of new card/review card you need to see. Once it's done, it finally loops over all deck and look for a card to learn.

This PR simply reverse the process. It split reset in two parts, which both can be used or ignored. A part which computes the number of cards to see and a part which reset all queues and list of decks to consider. When reset should be done in `getCard`, it first runs the important part of `reset` (i.e. resetting queues), then look for a card, then finally, in background, computes the numbers of cards.

Ideally, I would just separate `reset` in a pair of functions. I prefer to stay close to upsteam, and so I don't do this and just add variables to control which part of the resetting process I want to deactivate.



This is again a part of https://github.com/ankidroid/Anki-Android/pull/6010

It has a commit in common with https://github.com/ankidroid/Anki-Android/pull/6026 and it contains https://github.com/ankidroid/Anki-Android/pull/6034. Without both of those commits, this PR could not work. Which also means that this PR will get smaller once both other PR are accepted.

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1572,6 +1577,15 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
+        try {
+            return new TaskData(lambda.call());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628)

Arthur-Milchior commented:
Given that CI works and the problem is only the RuntimeException, I won't change this PR unless you know what else can be done to avoid it

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

Arthur-Milchior commented:
Rebased onto master because one of its commit was merged, which let to a conflict

Arthur-Milchior commented:
Rebased on master, because of merge conflict

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1562,6 +1567,15 @@ public TaskData doInBackgroundCheckCardSelection(TaskData... params) {
+        try {
+            return new TaskData(new Object[]{lambda.call()});
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628)

timrae replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid throwing raw exception types.](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628)

Arthur-Milchior commented:
the tests works on my phone. I have no idea what's wrong with the tests here specifically

> Task :AnkiDroid:testDebugUnitTest
com.ichi2.anki.AbstractFlashcardViewerTest > testTypeAnsAnswerFilterNormalEmpty FAILED
    Unable to resolve artifact: Missing:
    ----------
    1) org.robolectric:android-all:jar:9-robolectric-4913185-2
      Try downloading the file manually from the project website.

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

timrae commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -59,6 +59,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.concurrent.Callable;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'java.util.concurrent.Callable'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

Arthur-Milchior commented:
I pushed a new code. That's the same idea behind it, but I fouud a way to make the code cleaner while  keeping it as fast

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

Arthur-Milchior commented:
CI failed due to 
> /home/travis/.jabba/bin/jabba does not appear to be a valid binary.
Check your Internet connection / proxy settings and try again.

I'll ignore this bug until CI/github is repaired

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

Arthur-Milchior commented:
Rebased to solve merge conflict

timrae commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=5373628&bid=17615420)

Arthur-Milchior commented:
Rebased to remove the generic task "RUNNABLE" as requested by @david-allison-1 in another PR. Instead I use a task whose goal is specifically to reset the counts.

Arthur-Milchior commented:
Rebased to use the change in https://github.com/ankidroid/Anki-Android/pulls?q=is%3Apr+author%3AArthur-Milchior+is%3Aclosed

Arthur-Milchior commented:
Conflict removed

Arthur-Milchior commented:
Conflict removed

Arthur-Milchior commented:
Merge conflict removed

mikehardy commented:
Careful - if you push a blank (for whatever reason) it auto-closes 😱

Arthur-Milchior commented:
No idea what occurred

Arthur-Milchior commented:
Draft until https://github.com/ankidroid/Anki-Android/pull/6635 is considered

Arthur-Milchior commented:
I didn't expect it would only take a few hours

Arthur-Milchior commented:
Draft because upstream test found bug here :(

Arthur-Milchior commented:
Undrafted because the bug aas been corrected. It was actually an interaction with another PR.

Arthur-Milchior commented:
Conflict removed

Arthur-Milchior commented:
Rebased and corrected. I believe it's the last slow things I was able to correct. Getting the first card without computing counts, in order to win time.

mikehardy commented:
Scanning through our aging PR stack as I contemplate 2.14, @Arthur-Milchior I assume this one is still valid and strongly desired? Looks like it needs a rebase again

Arthur-Milchior commented:
This is still valid. But it still have the problem to have test errors that are related to background task time conflict. So probably depends on being able to run tests in foreground, which itself depends on having a TaskManager. So I don't expect to have it merged soon.
To be more specific, even if we decide that running all test in foreground is a bad idea, I still believe it would be nice to be able to run some of them without using background.

 https://github.com/ankidroid/Anki-Android/pull/7127 will conflict and is probably simpler to consider first.

Arthur-Milchior commented:
![image](https://user-images.githubusercontent.com/357361/94788899-8983d300-03d4-11eb-8253-38cf14eb3ee5.png)
Yeah, as I expected, it fails intermitently. 
There is a loop where each occurence call a background process and that does not wait properly for the process to end. That's not even a good simulation of actual use of ankidroid, since we never ask `getCard` to give us 10 cards immediately one after the other without waiting for the previous card to be computed and shown to the user

mikehardy commented:
failed unit tests in CI

Arthur-Milchior commented:
It works on computer. I fear I've still not added enough `advanceRobolectricLooperWithSleep`. Because here the problem really comes from the fact that a background task didn't have enough time to ends. Which would be impossible in real use. I try adding more of them. Hopefully, at some point I'll be able to use task manager to avoid this problem

mikehardy commented:
Well, we don't have Travis right now, and when I pull this branch locally, rebase it against tip of main branch and run it, everything passes 🤷  - we go!

mikehardy commented:
Thanks for your patience on this one - you are very understanding (which I am thankful for) and I just want to say that it is still on me to honor your work by reviewing it and merging it whenever possible. Your contributions are helpful and I even though you are understanding I am sad it has taken so long to get to them. But there is always hope for the future and while we are early in the 2.15 cycle I'm going to take another pass at moving all of them forward. Cheers Arthur

Arthur-Milchior commented:
I am very thankful for your to merge it. But, and it's a HUGE but.

This never worked on travis. It have worked on computer for quite some time, but I've never been able to find a way to have it work on travis, without using an alternate CollectionTask that sends nothing to background.
This mean that if we ever goes back to travis or anything similar, the test may fail.

mikehardy commented:
I just fixed the reason it wasn't working ;-), check the linked commit here that just went in
I was able to get it to "reliably" fail, maybe 20% of times locally, which was enough to isolate things
After that commit I'm 100% stable

Arthur-Milchior commented:
Oh.

I always thought it was the call below to `note()` that failed. That makes so much sens now. Of course all my tries would fail. thanks a lot !
